proj/hps_accel/gateware/mem.py

```diff
@@ -0,0 +1,75 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from nmigen import Cat, Memory, Module, Signal
+from nmigen_cfu.util import SimpleElaboratable
+
+
+class WideReadMemory(SimpleElaboratable):
+    """A memory that may be read 128 bits per cycle.
+
+    Writes to the memory are with 32 bit words, while read accesses read
+    four words concurrently.
+
+    Only a read or a write may take place on any cycle. In the event of
+    simultaneous writes and reads, the write takes precedence.
+
+    Parameters
+    ----------
+    depth: int
+      The number of words. Must be a positive multiple of four.
+
+    Attributes
+    ----------
+    write_addr: Signal(range(depth)), in
+      The address to which to write
+    write_data: Signal(32), in
+      The data to write
+    write_enable: Signal(), in
+      Write enable when set then data is written to the memory.
+    read_addr: Signal(range(depth / 4)), in
+      The address to read four words from. Lower two bits of words address
+      are not present since all four words are retrieved.
+    read_data: Signal(128), out
+      The four 32 bit words read from memory. Data is available the cycle
+      after the read_addr is placed on the bus
+    """
+
+    def __init__(self, depth):
+        self.__depth = depth
```

Comment:
Why the double-underscored attribute name?

I guess you probably know that single-underscore is conventional for private attributes, double-underscore has the surprising behaviour with name mangling.

Doesn't really matter either way in this case though since we don't expect this to be subclassed.

Reply:
Oops. That is showing my old-school Python roots.